internal/cache/cache.go

```diff
@@ -0,0 +1,199 @@
+package cache
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/buildkite/agent/v3/logger"
+	"github.com/buildkite/agent/v3/version"
+	"github.com/buildkite/zstash"
+	"github.com/buildkite/zstash/api"
+	"github.com/buildkite/zstash/cache"
+	"github.com/dustin/go-humanize"
+	"gopkg.in/yaml.v3"
+)
+
+// Config holds the configuration for cache operations
+type Config struct {
+	// BucketURL is the URL of the bucket (e.g., s3://bucket-name)
+	BucketURL string
+	// Branch is the branch associated with the cache
+	Branch string
+	// Pipeline is the pipeline slug for this cache
+	Pipeline string
+	// Organization is the organization slug for this cache
+	Organization string
+	// CacheConfigFile is the path to the cache configuration YAML file
+	CacheConfigFile string
+	// Ids is a comma-separated list of cache IDs (if empty, processes all caches)
+	Ids string
+	// APIEndpoint is the Agent API endpoint
+	APIEndpoint string
+	// APIToken is the access token used to authenticate
+	APIToken string
+}
+
+// CacheClient defines the interface for cache operations
+type CacheClient interface {
+	Save(ctx context.Context, cacheID string) (zstash.SaveResult, error)
+	Restore(ctx context.Context, cacheID string) (zstash.RestoreResult, error)
+	ListCaches() []cache.Cache
+}
+
+// Save saves caches based on the provided configuration and logs results as each cache is processed
+func Save(ctx context.Context, l logger.Logger, cfg Config) error {
+	cacheClient, cacheIDs, err := setupCacheClient(ctx, l, cfg)
+	if err != nil {
+		return err
+	}
+
+	if cacheClient == nil {
+		l.Info("No caches defined in the cache configuration file, nothing to save")
+		return nil
+	}
+
+	return saveWithClient(ctx, l, cacheClient, cacheIDs)
+}
+
+// Restore restores caches based on the provided configuration and logs results as each cache is processed
+func Restore(ctx context.Context, l logger.Logger, cfg Config) error {
+	cacheClient, cacheIDs, err := setupCacheClient(ctx, l, cfg)
+	if err != nil {
+		return err
+	}
+
+	if cacheClient == nil {
+		l.Info("No caches defined in the cache configuration file, nothing to restore")
+		return nil
+	}
+
+	return restoreWithClient(ctx, l, cacheClient, cacheIDs)
+}
+
+// loadCacheConfiguration loads cache configuration from a YAML file
+func loadCacheConfiguration(cacheConfigFile string) ([]cache.Cache, error) {
+	data, err := os.ReadFile(cacheConfigFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read cache config file: %w", err)
+	}
+
+	var caches []cache.Cache
+	if err := yaml.Unmarshal(data, &caches); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal cache config file: %w", err)
+	}
+
+	return caches, nil
+}
+
+// setupCacheClient creates a cache client and determines which cache IDs to process
+func setupCacheClient(ctx context.Context, l logger.Logger, cfg Config) (*zstash.Cache, []string, error) {
```

Comment:
[non-blocking] as we will eventually introduce more ways to specify cache configuration, I am not sure if we should use cache id as the first level concept here. Maybe we should turn it into a concrete `CacheConfig` object  and pass that around.

Reply:
Good point, will add validation of cache ids.

Reply:
@zhming0 sorry commented on the wrong change here, cache ids are used to selectively restore caches in a step. The common example is a rails app wiht ruby and node deps, being able to only restore node or ruby deps or both in a step is useful. 

This model follows other ci providers in their structure. 

A good example of a bigger cache is something we experimented here https://github.com/bk-playground/kitesocial/blob/main/.buildkite/cache.hosted.yml#L1-L9 which has a docker cache.